src/Microsoft.ComponentDetection.Detectors/dotnet/DotNetComponentDetector.cs

```diff
@@ -0,0 +1,186 @@
+    private readonly ICommandLineInvocationService commandLineInvocationService;
+    private readonly IDirectoryUtilityService directoryUtilityService;
+    private readonly IFileUtilityService fileUtilityService;
+    private readonly IPathUtilityService pathUtilityService;
+    private readonly LockFileFormat lockFileFormat = new();
+    private readonly Dictionary<string, string?> sdkVersionCache = [];
+    private string? sourceDirectory;
+    private string? sourceFileRootDirectory;
+
+    public DotNetComponentDetector(
+        IComponentStreamEnumerableFactory componentStreamEnumerableFactory,
+        ICommandLineInvocationService commandLineInvocationService,
+        IDirectoryUtilityService directoryUtilityService,
+        IFileUtilityService fileUtilityService,
+        IPathUtilityService pathUtilityService,
+        IObservableDirectoryWalkerFactory walkerFactory,
+        ILogger<DotNetComponentDetector> logger)
+    {
+        this.commandLineInvocationService = commandLineInvocationService;
+        this.ComponentStreamEnumerableFactory = componentStreamEnumerableFactory;
+        this.directoryUtilityService = directoryUtilityService;
+        this.fileUtilityService = fileUtilityService;
+        this.pathUtilityService = pathUtilityService;
+        this.Scanner = walkerFactory;
+        this.Logger = logger;
+    }
+
+    public override string Id => "DotNet";
+
+    public override IList<string> SearchPatterns { get; } = [LockFileFormat.AssetsFileName];
+
+    public override IEnumerable<ComponentType> SupportedComponentTypes => [ComponentType.DotNet];
+
+    public override int Version { get; } = 1;
+
+    public override IEnumerable<string> Categories => ["DotNet"];
+
+    private async Task<string?> RunDotNetVersionAsync(string workingDirectoryPath, CancellationToken cancellationToken)
+    {
+        var workingDirectory = new DirectoryInfo(workingDirectoryPath);
+
+        var process = await this.commandLineInvocationService.ExecuteCommandAsync("dotnet", ["dotnet.exe"], workingDirectory, cancellationToken, "--version").ConfigureAwait(false);
+        return process.ExitCode == 0 ? process.StdOut.Trim() : null;
+    }
+
+    public override Task<IndividualDetectorScanResult> ExecuteDetectorAsync(ScanRequest request, CancellationToken cancellationToken = default)
+    {
+        this.sourceDirectory = this.pathUtilityService.NormalizePath(request.SourceDirectory.FullName);
+        this.sourceFileRootDirectory = this.pathUtilityService.NormalizePath(request.SourceFileRoot?.FullName);
+
+        return base.ExecuteDetectorAsync(request, cancellationToken);
+    }
+
+    protected override async Task OnFileFoundAsync(ProcessRequest processRequest, IDictionary<string, string> detectorArgs, CancellationToken cancellationToken = default)
+    {
+        var lockFile = this.lockFileFormat.Read(processRequest.ComponentStream.Stream, processRequest.ComponentStream.Location);
```

Comment:
if lockfile you execute this, if global.json file, you run execute `dotnet --version` or fallback you would parse the json directly ignoring the `rollforward` piece.

Reply:
As I mentioned above I think it's more accurate to start all detection from the assets files. Let's discuss if we prefer `dotnet --version` even for global.json below.